File: frontend/src/app/core/services/registration-completeness.service.ts
import { Injectable } from '@angular/core';
import { Patient } from '../models/patient.model';
import { Coverage } from '../models/coverage.model';
import { PatientConsent } from '../models/coverage.model';
import { RegistrationCompleteness, MissingField, RegistrationRequirement } from '../models/registration-completeness.model';

@Injectable({
  providedIn: 'root'
})
export class RegistrationCompletenessService {
  private registrationRequirements: RegistrationRequirement = {
    demographicsRequired: [
      'legalName', // firstName + lastName
      'dateOfBirth',
      'birthSex', // sex field
      'phone', // phoneNumber
      'address1', // addressLine1
      'city',
      'state',
      'zip' // pincode or zipCode
    ],
    coverageRequired: [
      'primaryInsurancePayer', // payer from primary coverage
      'memberId' // memberId from primary coverage
    ],
    consentRequired: true
  };

  checkCompleteness(
    patient: Patient,
    coverage?: Coverage | null,
    consent?: PatientConsent | null
  ): RegistrationCompleteness {
    const missing: MissingField[] = [];
    
    // Check Demographics
    const legalName = this.getLegalName(patient);
    if (!legalName) {
      missing.push({
        key: 'legalName',
        field: 'legalName',
        label: 'Legal Name (First & Last)',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    if (!patient.dateOfBirth) {
      missing.push({
        key: 'dateOfBirth',
        field: 'dateOfBirth',
        label: 'Date of Birth',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    if (!patient.sex && !patient.gender) {
      missing.push({
        key: 'birthSex',
        field: 'birthSex',
        label: 'Birth Sex',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    if (!patient.phoneNumber && !patient.phone) {
      missing.push({
        key: 'phone',
        field: 'phone',
        label: 'Phone Number',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    if (!patient.addressLine1 && !patient.address) {
      missing.push({
        key: 'address1',
        field: 'address1',
        label: 'Address Line 1',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    if (!patient.city) {
      missing.push({
        key: 'city',
        field: 'city',
        label: 'City',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    if (!patient.state) {
      missing.push({
        key: 'state',
        field: 'state',
        label: 'State',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    if (!patient.pincode && !patient.zipCode) {
      missing.push({
        key: 'zip',
        field: 'zip',
        label: 'ZIP Code',
        severity: 'CRITICAL',
        section: 'Demographics'
      });
    }

    // Check email (WARN)
    if (!patient.emailAddress && !patient.email) {
      missing.push({
        key: 'email',
        field: 'email',
        label: 'Email Address',
        severity: 'WARN',
        section: 'Demographics'
      });
    }

    // Check Coverage
    if (!coverage || !coverage.payer) {
      missing.push({
        key: 'primaryInsurancePayer',
        field: 'primaryInsurancePayer',
        label: 'Primary Insurance Payer',
        severity: 'WARN', // Can be configured to CRITICAL
        section: 'Coverage'
      });
    }

    if (!coverage || !coverage.memberId) {
      missing.push({
        key: 'memberId',
        field: 'memberId',
        label: 'Insurance Member ID',
        severity: 'WARN', // Can be configured to CRITICAL
        section: 'Coverage'
      });
    }

    if (coverage) {
      if (coverage.eligibilityStatus === 'NOT_VERIFIED') {
        missing.push({
          key: 'eligibilityStatus',
          field: 'eligibilityStatus',
          label: 'Insurance Eligibility (Not Verified)',
          severity: 'WARN',
          section: 'Coverage'
        });
      } else if (coverage.eligibilityStatus === 'EXPIRED' || coverage.eligibilityStatus === 'INACTIVE') {
        missing.push({
          key: 'eligibilityStatus',
          field: 'eligibilityStatus',
          label: `Insurance Eligibility (${coverage.eligibilityStatus})`,
          severity: 'CRITICAL',
          section: 'Coverage'
        });
      }
    }

    // Check Consent
    if (this.registrationRequirements.consentRequired) {
      if (!consent || !consent.consentSigned) {
        missing.push({
          key: 'consentSigned',
          field: 'consentSigned',
          label: 'Consent Not Signed',
          severity: 'CRITICAL',
          section: 'Consent'
        });
      }
    }

    const blockers = missing.filter(m => m.severity === 'CRITICAL');
    const warnings = missing.filter(m => m.severity === 'WARN');

    // Calculate completeness percentage
    const totalFields = 
      this.registrationRequirements.demographicsRequired.length +
      this.registrationRequirements.coverageRequired.length +
      (this.registrationRequirements.consentRequired ? 1 : 0);
    
    const completedFields = totalFields - missing.length;
    const completenessPct = totalFields > 0 
      ? Math.round((completedFields / totalFields) * 100) 
      : 100;

    return {
      completenessPct,
      missing,
      blockers,
      warnings
    };
  }

  private getLegalName(patient: Patient): string {
    if (patient.firstName && patient.lastName) {
      return `${patient.firstName} ${patient.lastName}`.trim();
    }
    return '';
  }

  getRequirements(): RegistrationRequirement {
    return { ...this.registrationRequirements };
  }

  updateRequirements(requirements: Partial<RegistrationRequirement>): void {
    this.registrationRequirements = {
      ...this.registrationRequirements,
      ...requirements
    };
  }
}

File: backend/src/main/java/com/ehr/staffservice/service/RegistrationCompletenessService.java
package com.ehr.staffservice.service;

import com.ehr.staffservice.dto.PatientProfileDto;
import com.ehr.staffservice.entity.Patient;
import com.ehr.staffservice.entity.PatientAddress;
import com.ehr.staffservice.repository.PatientAddressRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

/**
 * Service to compute patient registration completeness.
 * Used by both Admin and MyChart to show registration status.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class RegistrationCompletenessService {

    private final PatientAddressRepository addressRepository;

    /**
     * Compute registration completeness for a patient.
     * Returns percentComplete (0-100) and list of missing fields with severity.
     * 
     * This matches the frontend registration completeness logic:
     * - Demographics: legalName, dateOfBirth, birthSex, phone, address1, city, state, zip (8 fields)
     * - Coverage: primaryInsurancePayer, memberId (2 fields)
     * - Total: 10 required fields (consent is handled separately in frontend)
     * - Calculation: (completedFields / totalFields) * 100
     */
    @Transactional(readOnly = true)
    public PatientProfileDto.RegistrationStatusDto computeRegistrationStatus(Patient patient) {
        List<PatientProfileDto.MissingFieldDto> missingFields = new ArrayList<>();
        
        // Get primary address if available, otherwise check patient table address fields
        List<PatientAddress> addresses = addressRepository.findByPatientIdAndIsPrimary(patient.getPatientId(), true);
        PatientAddress primaryAddress = addresses.isEmpty() ? null : addresses.get(0);
        
        // Check DEMOGRAPHICS group (matches frontend logic)
        // 1. Legal Name (First + Last)
        if (patient.getFirstName() == null || patient.getFirstName().trim().isEmpty() ||
            patient.getLastName() == null || patient.getLastName().trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "legalName", "Legal Name (First & Last)", "CRITICAL"));
        }
        
        // 2. Date of Birth
        if (patient.getDateOfBirth() == null) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "dateOfBirth", "Date of Birth", "CRITICAL"));
        }
        
        // 3. Birth Sex (gender field)
        if (patient.getGender() == null || patient.getGender().trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "birthSex", "Birth Sex", "CRITICAL"));
        }
        
        // 4. Phone Number
        if (patient.getPhoneNumber() == null || patient.getPhoneNumber().trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "phone", "Phone Number", "CRITICAL"));
        }
        
        // 5-8. Address fields (check primaryAddress first, then fallback to patient table)
        String addressLine1 = null;
        String city = null;
        String state = null;
        String zipCode = null;
        
        if (primaryAddress != null) {
            addressLine1 = primaryAddress.getAddressLine1();
            city = primaryAddress.getCity();
            state = primaryAddress.getStateProvince();
            zipCode = primaryAddress.getPostalCode();
        } else {
            // Fallback to patient table address fields
            addressLine1 = patient.getAddressLine1();
            city = patient.getCity();
            state = patient.getState();
            zipCode = patient.getZipCode();
        }
        
        if (addressLine1 == null || addressLine1.trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "address1", "Address Line 1", "CRITICAL"));
        }
        if (city == null || city.trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "city", "City", "CRITICAL"));
        }
        if (state == null || state.trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "state", "State", "CRITICAL"));
        }
        if (zipCode == null || zipCode.trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "zip", "ZIP Code", "CRITICAL"));
        }
        
        // Check email (WARNING level, not counted in completeness)
        if (patient.getEmail() == null || patient.getEmail().trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "DEMOGRAPHICS", "email", "Email Address", "WARNING"));
        }
        
        // Check COVERAGE group
        if (patient.getInsuranceProvider() == null || patient.getInsuranceProvider().trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "COVERAGE", "primaryInsurancePayer", "Primary Insurance Payer", "WARNING"));
        }
        if (patient.getInsurancePolicyNumber() == null || patient.getInsurancePolicyNumber().trim().isEmpty()) {
            missingFields.add(new PatientProfileDto.MissingFieldDto(
                    "COVERAGE", "memberId", "Insurance Member ID", "WARNING"));
        }
        
        // Calculate percent complete (matches frontend calculation)
        // Total required fields: 8 demographics (CRITICAL) + 2 coverage (WARNING) = 10 fields
        // Only count CRITICAL and WARNING fields in completeness calculation
        int totalRequiredFields = 10; // 8 demographics + 2 coverage
        long missingRequiredFields = missingFields.stream()
                .filter(f -> "CRITICAL".equals(f.getSeverity()) || "WARNING".equals(f.getSeverity()))
                .count();
        
        int completedFields = totalRequiredFields - (int)missingRequiredFields;
        int percentComplete = totalRequiredFields > 0 
                ? Math.round((completedFields * 100) / totalRequiredFields)
                : 100;
        
        // Ensure percentComplete is between 0 and 100
        percentComplete = Math.max(0, Math.min(100, percentComplete));
        
        return new PatientProfileDto.RegistrationStatusDto(percentComplete, missingFields);
    }
}

